app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,175 @@
+                Log.v(TAG, "updateFromConfig: Add device '" + device.deviceID + "' to cache model");
+                deviceFolderMap.put(device.deviceID, defaultDevice);
+            }
+        }
+
+        // Handle folders that were removed from the config.
+        List<String> removedFolders = new ArrayList<>();;
+        Boolean folderFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            for (String folderId : deviceFolderMap.get(deviceId).keySet()) {
+                folderFound = false;
+                for (Folder folder : newFolderSet) {
+                    if (folder.id.equals(folderId)) {
+                        folderFound = true;
+                        break;
+                    }
+                }
+                if (!folderFound) {
+                    removedFolders.add(folderId);
+                }
+            }
+        }
+        for (String folderId : removedFolders) {
+            Log.v(TAG, "updateFromConfig: Remove folder '" + folderId + "' from cache model");
+            removeFolder(folderId);
+        }
+
+        // Handle folders that were added to the config.
+        for (Folder folder : newFolderSet) {
+            for (Device device : newDeviceSet) {
+                if (folder.getDevice(device.deviceID) != null) {
+                    // folder is shared with device.
+                    if (!deviceFolderMap.get(device.deviceID).containsKey(folder.id)) {
+                        Log.v(TAG, "updateFromConfig: Add folder '" + folder.id +
+                                    "' shared with device '" + device.deviceID + "' to cache model.");
+                        deviceFolderMap.get(device.deviceID).put(folder.id, defaultFolder);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Calculates remote device sync completion percentage across all folders
+     * shared with the device.
+     */
+    public int getDeviceCompletion(String deviceId) {
+        int folderCount = 0;
+        double sumCompletion = 0;
+        for (String folderId : deviceFolderMap.getOrDefault(deviceId, defaultDevice).keySet()) {
+            sumCompletion += (deviceFolderMap.get(deviceId)).get(folderId).completion;
+            folderCount++;
+        }
+        if (folderCount == 0) {
+            return 100;
+        } else {
+            return (int) Math.floor(sumCompletion / folderCount);
+        }
+    }
+
+    /**
+     * Set completionInfo within the completion[deviceId][folderId] model.
+     */
+    public void setCompletionInfo(String deviceId, String folderId,
+                                    CompletionInfo completionInfo) {
+        // Add device parent node if it does not exist.
+        if (!deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.put(deviceId, defaultDevice);
+        }
+        // Add folder or update existing folder entry.
+        deviceFolderMap.get(deviceId).put(folderId, completionInfo);
+    }

```

Comment:
Again, the default value is misused, perhaps do:

```
if deviceMap.contains(deviceId) {
  deviceMap.get(deviceId).put(folderId, completionInfo)
} else {
  Map<String, Completion> deviceCompletion = new HashMap<String, Completion>()
  deviceCompletion.put(folder, completion)
  deviceMap.put(device, deviceCompeltion)
}
```

Reply:
fixed, see above

Reply:
oh wait, in the mentioned lines is no getOrDefault nor is there a defaultFolder or deviceDevice ?